modules/password/password.go

```diff
@@ -0,0 +1,58 @@
+package password
+
+import (
+	"math/rand"
+	"regexp"
+	"sync"
+	"time"
+
+	"code.gitea.io/gitea/modules/setting"
+)
+
+var matchComplexities = map[string]regexp.Regexp{}
+var matchComplexityOnce sync.Once
+
+// CheckPasswordComplexity return True if password is Complexity
+func CheckPasswordComplexity(pwd string) bool {
+	if len(setting.PasswordComplexity) > 0 {
+		matchComplexityOnce.Do(func() {
+			for key, val := range setting.PasswordComplexity {
+				matchComplexity := regexp.MustCompile(val)
+				matchComplexities[key] = *matchComplexity
+			}
+		})
+		for _, val := range matchComplexities {
+			if !val.MatchString(pwd) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// Generate  a random password
+func Generate(n int) string {
+	rand.Seed(time.Now().UnixNano())
+	var dict = map[int]string{
+		0: "abcdefghijklmnopqrstuvwxyz",
+		1: "ABCDEFGHIJKLMNOPQRSTUVWXYZ",
+		2: "0123456789",
+		3: "_-",
+	}
+	buffer := make([]byte, n)
```

Comment:
You can prepare a string that contains the sum of all the enabled character classes and randomize from there. For example:
```
enabledCharacters := ""
for j := 0; j < len(dict); j++ {
	if thisComplexityEnabled(j) { // Function is just an example
		enabledCharacters = enabledCharacters+ dict[j]
	}
}
// Now enabledCharacters may contain:
// "abcdefghijklmnopqrstuvwxyz"
// "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
// "abcdefghijklmnopqrstuvwxyz0123456789_-"
// etc.
```

Then in your generation loop below you don't use `dict` at all, but `enabledCharacters`, which contains all the needed characters.

In fact, you don't need a dictionary for this. A simple slice will suffice.

Reply:
Ok,  I will change the code